jni/src/faiss_wrapper.cpp

```diff
@@ -0,0 +1,347 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+#include "jni_util.h"
+#include "faiss_wrapper.h"
+
+#include "faiss/index_factory.h"
+#include "faiss/MetaIndexes.h"
+#include "faiss/index_io.h"
+#include "faiss/IndexHNSW.h"
+#include "faiss/IndexIVFFlat.h"
+
+#include <algorithm>
+#include <jni.h>
+#include <vector>
+#include <string>
+#include <faiss/impl/io.h>
+
+
+// Translate space type to faiss metric
+faiss::MetricType TranslateSpaceToMetric(const std::string& spaceType);
+
+// Set additional parameters on faiss index
+void SetExtraParameters(JNIEnv *env, const std::unordered_map<std::string, jobject>& parametersCpp,
+                        faiss::Index * index);
+
+// Train an index with data provided
+void InternalTrainIndex(faiss::Index * index, faiss::Index::idx_t n, const float* x);
+
+void knn_jni::faiss_wrapper::CreateIndex(JNIEnv * env, jintArray idsJ, jobjectArray vectorsJ, jstring indexPathJ,
+                                         jobject parametersJ) {
+
+    if (idsJ == nullptr) {
+        throw std::runtime_error("IDs cannot be null");
+    }
+
+    if (vectorsJ == nullptr) {
+        throw std::runtime_error("Vectors cannot be null");
+    }
+
+    if (indexPathJ == nullptr) {
+        throw std::runtime_error("Index path cannot be null");
+    }
+
+    if (parametersJ == nullptr) {
+        throw std::runtime_error("Parameters cannot be null");
+    }
+
```

Comment:
Will these exceptions crash the jvm as these are cpp Exceptions. My understanding Java does not understand these exceptions and crash? If this crashes JVM, then we may need to type case to Java Exception. Same comment for all other places where we throw cpp Exceptions.

Reply:
We wrap all of the calls to wrappers in https://github.com/jmazanec15/k-NN-1/blob/faiss-initial/jni/src/org_opensearch_knn_index_JNIService.cpp#L20. So, this will ensure that exception is handled correctly. Additionally, some of the tests test exceptions.